File: day-19/monster_messages.py
import re

input_file = 'day-19/input.txt'

def parse_rules_section(rule_strings):
  rules = {}
  for rule in rule_strings.splitlines():
    num, rule_string = rule.split(':', 1)
    if '"' in rule_string:
      rules[int(num)] = rule_string.strip().replace('"','')
    else:
      rules[int(num)] = [[int(s) for s in rule_split.split()] for rule_split in rule_string.split('|')]
  return rules

def resolve_rule(rules, rule_no, part2):
  rule = rules[rule_no]
  if isinstance(rule, str):
    return rule
  if part2:
    if rule_no == 8:
      return resolve_rule(rules, 42, part2) + '+'
    elif rule_no == 11:
      rule_42 = resolve_rule(rules, 42, part2)
      rule_31 = resolve_rule(rules, 31, part2)
      repeat_matches = []
      for i in range(1,5):
        repeat_matches.append(rule_42+'{'+str(i) +'}' + rule_31 + '{'+str(i) +'}')
      return '(?:' + '|'.join(repeat_matches) + ')'
  if len(rule) == 1 and len(rule[0]) == 1 and isinstance(rule[0], str):
    rules[rule_no] = rule[0]
  else:
    resolved_parts = []
    for or_rules in rule:
      resolved = []
      for linked_rule in or_rules:
        resolved.append(resolve_rule(rules, linked_rule, part2))
      resolved_parts.append(''.join(resolved))
    rules[rule_no] = '(?:' + '|'.join(resolved_parts) + ')'
  return rules[rule_no]

def count_matches(input, part2 = False):
  sections = input.split('\n\n')

  rules_dict = parse_rules_section(sections[0])
  rule_0 = resolve_rule(rules_dict, 0, part2)
  
  return sum(1 for message in sections[1].splitlines() if re.fullmatch(rule_0, message) != None)

def part1(input):
  return count_matches(input)
  
def part2(input):
  return count_matches(input, True)

if __name__ == "__main__":
  with open(input_file) as f:
    data = f.read()
  print("Part 1: ", part1(data))
  print("Part 2: ", part2(data))

File: day-19/monster_messages.go
package main

import (
	"fmt"
	"io/ioutil"
	"regexp"
	"strconv"
	"strings"
)

const inputFile string = "./day-19/input.txt"

type Rule struct {
	processed string
	subRules  [][]int
}

func ParseRules(rulesSection string) map[int]Rule {
	rules := make(map[int]Rule)
	for _, rule := range strings.Split(strings.TrimSpace(rulesSection), "\n") {
		split := strings.SplitN(rule, ": ", 2)
		ruleContents := split[1]
		rule := Rule{}
		if strings.Contains(ruleContents, "\"") {
			rule.processed = strings.Trim(strings.TrimSpace(ruleContents), "\"")
		} else {
			orMatchers := [][]int{}
			for _, ruleSplit := range strings.Split(ruleContents, "|") {
				splitMatcher := []int{}
				for _, numString := range strings.Split(strings.TrimSpace(ruleSplit), " ") {
					num, _ := strconv.Atoi(numString)
					splitMatcher = append(splitMatcher, num)
				}
				orMatchers = append(orMatchers, splitMatcher)
			}
			rule.subRules = orMatchers
		}

		ruleNum, _ := strconv.Atoi(split[0])
		rules[ruleNum] = rule
	}
	return rules
}

//TODO - put function on rule?
func ResolveRule(rules map[int]Rule, ruleNum int, part2 bool) string {
	rule := rules[ruleNum]
	if rule.processed != "" {
		return rule.processed
	}
	if part2 {
		if ruleNum == 8 {
			return ResolveRule(rules, 42, part2) + "+"
		} else if ruleNum == 11 {
			rule42 := ResolveRule(rules, 42, part2)
			rule31 := ResolveRule(rules, 31, part2)
			repeatMatches := []string{}
			for i := 1; i <= 5; i++ {
				repeatMatches = append(repeatMatches, fmt.Sprintf("%s{%d}%s{%d}", rule42, i, rule31, i))
			}
			rule.processed = "(?:" + strings.Join(repeatMatches, "|") + ")"
			return rule.processed
		}
	}
	resolvedParts := []string{}
	for _, orRules := range rule.subRules {
		resolved := []string{}
		for _, linkedRule := range orRules {
			resolved = append(resolved, ResolveRule(rules, linkedRule, part2))
		}
		resolvedParts = append(resolvedParts, strings.Join(resolved, ""))
	}
	rule.processed = "(?:" + strings.Join(resolvedParts, "|") + ")"
	return rule.processed
}

func CountMatches(input string, part2 bool) int {
	sections := strings.Split(input, "\n\n")
	rules := ParseRules(sections[0])
	ruleZero := regexp.MustCompile("^" + ResolveRule(rules, 0, part2) + "$")

	matches := 0
	for _, message := range strings.Split(sections[1], "\n") {
		if ruleZero.MatchString(message) {
			matches++
		}
	}
	return matches
}

func Part1(input string) int {
	return CountMatches(input, false)
}

func Part2(input string) int {
	return CountMatches(input, true)
}

func main() {
	b, err := ioutil.ReadFile(inputFile)
	if err != nil {
		fmt.Print(err)
	}
	data := string(b)

	fmt.Printf("Part 1: %d\n", Part1(data))
	fmt.Printf("Part 2: %d\n", Part2(data))
}
